File: src/main/javabak/gui/widgets/ProgressBarWidget.java
package redcrafter07.processed.gui.widgets;

import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.Renderable;
import net.minecraft.resources.ResourceLocation;
import java.util.function.Supplier;

public class ProgressBarWidget implements Renderable {
    private final int x;
    private final int y;
    private final ProgressBarData data;
    private Supplier<Double> progressSupplier;
    public ProgressBarWidget(int x, int y, ProgressBarData data, Supplier<Double> progressSupplier) {
        this.x = x;
        this.y = y;
        this.data = data;
        this.progressSupplier = progressSupplier;
    }

    private void blit(GuiGraphics graphics, int x, int y, int offX, int offY, int width, int height) {
        graphics.blit(data.texture, x, y, offX, offY, width, height, data.width * 2, data.height);
    }

    @Override
    public void render(GuiGraphics graphics, int mouseX, int mouseY, float partialTick) {
        blit(graphics, x, y, 0, 0, data.width, data.height);

        if (data.direction == ProgressBarDirection.Down) {
            final int height = (int) (progressSupplier.get() * data.height);
            blit(graphics, x, y, data.width, 0, data.width, height);
            return;
        }
        if (data.direction == ProgressBarDirection.Left) {
            final int width = (int) (progressSupplier.get() * data.width);
            blit(graphics, x, y, data.width, 0, width, data.height);
        }

        if (data.direction == ProgressBarDirection.Up) {
            final int height = (int) (progressSupplier.get() * data.height);
            blit(graphics, x, y + data.height - height, data.width, data.height - height, data.width, height);
            return;
        }
        if (data.direction == ProgressBarDirection.Right) {
            final int width = (int) (progressSupplier.get() * data.width);
            blit(graphics, x + data.width - width, y, data.width * 2 - width, 0, width, data.height);
        }
    }

    public void setProgressSupplier(Supplier<Double> newSupplier) {
        progressSupplier = newSupplier;
    }

    public enum ProgressBarDirection {
        Down,
        Up,
        Right,
        Left
    }

    /**
     * @param width This should be *half* the width of the file, as it only matches the width of the "off-state" of the progress bar. However, the file has both the on and off state and thus is double the width of one of the states
     */
    public record ProgressBarData(
            ResourceLocation texture,
        int width,
        int height,
            ProgressBarDirection direction
    ) {
        public ProgressBarWidget create(int x, int y, Supplier<Double> supplier) {
            return new ProgressBarWidget(x, y, this, supplier);
        }
    }
}

File: src/main/kotlin/redcrafter07/processed/gui/widgets/ProgressBarWidget.kt
package redcrafter07.processed.gui.widgets

import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.components.Renderable
import net.minecraft.client.gui.components.events.GuiEventListener
import net.minecraft.client.gui.narration.NarratableEntry
import net.minecraft.client.gui.narration.NarrationElementOutput
import net.minecraft.resources.ResourceLocation
import net.neoforged.fml.ModList
import java.util.function.Supplier

class ProgressBarWidget(
    val x: Int, val y: Int, val data: ProgressBarData, val progressSupplier: Supplier<Double>
) : Renderable, GuiEventListener, NarratableEntry {
    private fun blit(graphics: GuiGraphics, x: Int, y: Int, offX: Int, offY: Int, width: Int, height: Int) {
        graphics.blit(data.texture, x, y, offX.toFloat(), offY.toFloat(), width, height, data.width * 2, data.height)
    }

    override fun render(graphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
        blit(graphics, x, y, 0, 0, data.width, data.height)

        when (data.direction) {
            ProgressBarDirection.Up -> {
                val height = (progressSupplier.get() * data.height).toInt()
                blit(graphics, x, y + data.height - height, data.width, data.height - height, data.width, height)
            }

            ProgressBarDirection.Down -> {
                val height = (progressSupplier.get() * data.height).toInt()
                blit(graphics, x, y, data.width, 0, data.width, height)
            }

            ProgressBarDirection.Left -> {
                val width = (progressSupplier.get() * data.width).toInt()
                blit(graphics, x + data.width - width, y, data.width * 2 - width, 0, width, data.height)
            }

            ProgressBarDirection.Right -> {
                val width = (progressSupplier.get() * data.width).toInt()
                blit(graphics, x, y, data.width, 0, width, data.height)
            }
        }
    }

    override fun setFocused(p0: Boolean) {
    }

    override fun isFocused(): Boolean = false
    override fun narrationPriority(): NarratableEntry.NarrationPriority = NarratableEntry.NarrationPriority.NONE
    override fun updateNarration(p0: NarrationElementOutput) {
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        val x = mouseX.toInt()
        val y = mouseY.toInt()
        if (button == 0 && x >= this.x && y >= this.y && x < this.x + data.width && y < this.y + data.height) {
            if (ModList.get().isLoaded("emi")) {
                // NOTE: Not sure if these full paths are necessary in order to not load the class when emi isn't loaded
                val cat = redcrafter07.processed.integration.emi.Categories.fromProgressBar(data)
                if (cat != null) {
                    dev.emi.emi.api.EmiApi.displayRecipeCategory(cat)
                    return true
                }

            }
        }
        return super.mouseClicked(mouseX, mouseY, button)
    }

    enum class ProgressBarDirection {
        Down, Up, Right, Left
    }

    /**
     * @param width This should be *half* the width of the file, as it only matches the width of the "off-state" of the progress bar. However, the file has both the on and off state and thus is double the width of one of the states
     */
    data class ProgressBarData(
        val texture: ResourceLocation, val width: Int, val height: Int, val direction: ProgressBarDirection, val id: String
    ) {
        fun create(x: Int, y: Int, supplier: Supplier<Double>): ProgressBarWidget {
            return ProgressBarWidget(x, y, this, supplier)
        }
    }
}